sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +211,281 @@ object Hash {
+    ): NodeBuilder =
+      NodeBuilder.LanguageVersionToNodeBuilder
+        .getOrElse(version, throw NodeHashingError.UnsupportedLanguageVersion(version))
+        .apply(messageDigest, valueBuilder)
+
+    private[lf] def builderForNode(
+        node: Node,
+        messageDigest: MessageDigestWithContext,
+        valueBuilder: ValueHashBuilder,
+    ): NodeBuilder = node.optVersion match {
+      case Some(version) => builderForVersion(version, messageDigest, valueBuilder)
+      // If the node is version agnostic, the builder doesn't matter so pick V1
+      // This only applies to rollback nodes
+      case None => new NodeBuilderV1(messageDigest, valueBuilder)
+    }
+  }
+
+  private final class NodeBuilderV1(
+      messageDigest: MessageDigestWithContext,
+      override protected val valueBuilder: ValueHashBuilder,
+  ) extends NodeBuilder(NodeHashVersion.V1, messageDigest, valueBuilder) {
+    private def addCreateNode(create: Node.Create): this.type = {
+      addContext("Create Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(create.coid, _.addCid))
+        .withContext("Package Name")(_.addString(create.packageName))
+        .withContext("Template Id")(_.addIdentifier(create.templateId))
+        .withContext("Arg")(_.addWithValueBuilder(create.arg, _.addTypedValue))
+        .withContext("Signatories")(_.addStringSet(create.signatories))
+        .withContext("Stakeholders")(_.addStringSet(create.stakeholders))
+        .withContext("Global Keys")(_.addOptional(create.keyOpt, _.addGlobalKeyWithMaintainers))
+    }
+
+    private def addFetchNode(fetch: Node.Fetch): NodeBuilder = {
+      addContext("Fetch Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(fetch.coid, _.addCid))
+        .withContext("Package Name")(_.addString(fetch.packageName))
+        .withContext("Template Id")(_.addIdentifier(fetch.templateId))
+        .withContext("Signatories")(_.addStringSet(fetch.signatories))
+        .withContext("Stakeholders")(_.addStringSet(fetch.stakeholders))
+        .withContext("Acting Parties")(_.addStringSet(fetch.actingParties))
+        .withContext("Global Keys")(_.addOptional(fetch.keyOpt, _.addGlobalKeyWithMaintainers))
+        .withContext("By Key")(_.addBool(fetch.byKey))
+        .withContext("Interface Id")(_.addOptional(fetch.interfaceId, _.addIdentifier))
+    }
+
+    private def addExerciseNode(exercise: Node.Exercise, nodes: Map[NodeId, Node]): NodeBuilder = {
+      addContext("Exercise Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(exercise.targetCoid, _.addCid))
+        .withContext("Package Name")(_.addString(exercise.packageName))
+        .withContext("Template Id")(_.addIdentifier(exercise.templateId))
+        .withContext("Signatories")(_.addStringSet(exercise.signatories))
+        .withContext("Stakeholders")(_.addStringSet(exercise.stakeholders))
+        .withContext("Acting Parties")(_.addStringSet(exercise.actingParties))
```

Comment:
Should we use the same convention to encode and hash the stackholders of create and exe ?
using `val non_signatory_stakeholders = stakeholders -- signatories` instead of `stackholder`.
@andreaslochbihler-da ?

Reply:
I think it makes sense to be consistent here. Signatories and non-signatory stakeholders make sense to me, but it's not a huge deal.